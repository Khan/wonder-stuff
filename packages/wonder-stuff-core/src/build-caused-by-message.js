// @flow
/**
 * Builds a string that describes the causal relationship between two messages.
 *
 * @param {string} consequence The first message that describes the consequence
 * that occurred.
 * @param {string} cause The second message that describes the cause.
 * @returns {string} A string that indicates `consequence` is caused by `cause`.
 */
export const buildCausedByMessage = (
    consequence: ?string,
    cause: ?string,
): string => {
    const messageOrDefault = (message: ?string): string =>
        message?.trim() || "(empty message)";

    const consequenceMessage = messageOrDefault(consequence);
    if (cause == null) {
        return consequenceMessage;
    }

    return `${consequenceMessage}\n\tcaused by\n\t\t${messageOrDefault(cause)}`;
};
