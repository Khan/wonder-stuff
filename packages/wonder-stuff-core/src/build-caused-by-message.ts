/**
 * Builds a string that describes the causal relationship between two messages.
 *
 * @param {string} consequence The first message that describes the consequence
 * that occurred.
 * @param {string} cause The second message that describes the cause.
 * @returns {string} A string that indicates `consequence` is caused by `cause`.
 */
export const buildCausedByMessage = (
    consequence?: string | null,
    cause?: string | null,
): string => {
    const messageOrDefault = (message?: string | null): string =>
        message?.trim() || "(empty message)";

    const consequenceMessage = messageOrDefault(consequence);
    if (cause == null) {
        return consequenceMessage;
    }

    return `${consequenceMessage}\n\tcaused by\n\t\t${messageOrDefault(cause)}`;
};
