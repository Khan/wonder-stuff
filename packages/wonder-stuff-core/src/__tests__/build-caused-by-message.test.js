// @flow
import {buildCausedByMessage} from "../build-caused-by-message.js";

describe("#buildCausedByMessage", () => {
    it("should combine the strings into a 'caused by' message", () => {
        // Arrange
        const consequence = "No more Halloween candy";
        const cause = "Gluttonous parents";

        // Act
        const result = buildCausedByMessage(consequence, cause);

        // Assert
        expect(result).toMatchInlineSnapshot(`
            "No more Halloween candy
            	caused by
            		Gluttonous parents"
        `);
    });

    it.each([null, undefined])(
        "should only include cause if it is non-null",
        (testCase) => {
            // Arrange
            const consequence = "No more Halloween candy";
            const cause = testCase;

            // Act
            const result = buildCausedByMessage(consequence, cause);

            // Assert
            expect(result).toBe(consequence);
        },
    );

    it.each([null, undefined])(
        "should default nullish consequences to (empty message)",
        (testCase) => {
            // Arrange
            const consequence = testCase;
            const cause = null;

            // Act
            const result = buildCausedByMessage(consequence, cause);

            // Assert
            expect(result).toBe("(empty message)");
        },
    );

    it.each(["    ", "\n\t\t"])(
        "should default whitespace strings to (empty message)",
        (testCase) => {
            // Arrange
            const consequence = testCase;
            const cause = testCase;

            // Act
            const result = buildCausedByMessage(consequence, cause);

            // Assert
            expect(result).toMatchSnapshot();
        },
    );
});
