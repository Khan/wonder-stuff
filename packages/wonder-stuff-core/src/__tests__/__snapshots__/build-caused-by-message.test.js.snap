// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildCausedByMessage should default whitespace strings to (empty message) 1`] = `
"(empty message)
	caused by
		(empty message)"
`;

exports[`#buildCausedByMessage should default whitespace strings to (empty message) 2`] = `
"(empty message)
	caused by
		(empty message)"
`;
